cmd/influx/write.go

```diff
@@ -68,94 +75,128 @@ or add an entire file specified with an @ prefix.`
 		}
-	}
-	if writeFlags.Bucket != "" {
-		filter.Name = &writeFlags.Bucket
-	}
 
-	if writeFlags.OrgID != "" {
-		filter.OrganizationID, err = platform.IDFromString(writeFlags.OrgID)
-		if err != nil {
-			return fmt.Errorf("failed to decode org-id id: %v", err)
+		var filter platform.BucketFilter
+		if writeFlags.BucketID != "" {
+			filter.ID, err = platform.IDFromString(writeFlags.BucketID)
+			if err != nil {
+				return fmt.Errorf("failed to decode bucket-id: %v", err)
+			}
+		}
+		if writeFlags.Bucket != "" {
+			filter.Name = &writeFlags.Bucket
 		}
-	}
-	if writeFlags.Org != "" {
-		filter.Org = &writeFlags.Org
-	}
 
-	buckets, n, err := bs.FindBuckets(ctx, filter)
-	if err != nil {
-		return fmt.Errorf("failed to retrieve buckets: %v", err)
-	}
+		if writeFlags.OrgID != "" {
+			filter.OrganizationID, err = platform.IDFromString(writeFlags.OrgID)
+			if err != nil {
+				return fmt.Errorf("failed to decode org-id id: %v", err)
+			}
+		}
+		if writeFlags.Org != "" {
+			filter.Org = &writeFlags.Org
+		}
 
-	if n == 0 {
-		if writeFlags.Bucket != "" {
-			return fmt.Errorf("bucket %q was not found", writeFlags.Bucket)
+		buckets, n, err := bs.FindBuckets(ctx, filter)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve buckets: %v", err)
 		}
 
-		if writeFlags.BucketID != "" {
-			return fmt.Errorf("bucket with id %q does not exist", writeFlags.BucketID)
+		if n == 0 {
+			if writeFlags.Bucket != "" {
+				return fmt.Errorf("bucket %q was not found", writeFlags.Bucket)
+			}
+
+			if writeFlags.BucketID != "" {
+				return fmt.Errorf("bucket with id %q does not exist", writeFlags.BucketID)
+			}
 		}
-	}
 
-	bucketID, orgID := buckets[0].ID, buckets[0].OrgID
+		bucketID, orgID = buckets[0].ID, buckets[0].OrgID
+	}
 
 	var r io.Reader
-	if args[0] == "-" {
-		r = os.Stdin
-	} else if len(args[0]) > 0 && args[0][0] == '@' {
-		f, err := os.Open(args[0][1:])
+	if len(args) > 0 && args[0][0] == '@' {
+		// backward compatibility
+		writeFlags.File = args[0][1:]
+	}
+
+	if len(writeFlags.File) > 0 {
+		f, err := os.Open(writeFlags.File)
 		if err != nil {
-			return fmt.Errorf("failed to open %q: %v", args[0][1:], err)
+			return fmt.Errorf("failed to open %q: %v", writeFlags.File, err)
 		}
 		defer f.Close()
 		r = f
+		if len(writeFlags.Format) == 0 && strings.HasSuffix(writeFlags.File, ".csv") {
+			writeFlags.Format = inputFormatCsv
+		}
+	} else if len(args) == 0 || args[0] == "-" {
+		r = os.Stdin
 	} else {
 		r = strings.NewReader(args[0])
 	}
+	// validate input format
+	if len(writeFlags.Format) > 0 && writeFlags.Format != inputFormatLineProtocol && writeFlags.Format != inputFormatCsv {
+		return fmt.Errorf("unsupported input format: %s", writeFlags.Format)
+	}
 
-	s := write.Batcher{
-		Service: &http.WriteService{
-			Addr:               flags.Host,
-			Token:              flags.Token,
-			Precision:          writeFlags.Precision,
-			InsecureSkipVerify: flags.skipVerify,
-		},
+	if writeFlags.Format == inputFormatCsv {
```

Comment:
can we inspect the filename to determine if it is `csv` vs `lp`? if so, we can make this a bit simpler for hte user by not requiring the format, and inspecting the file instead.

Reply:
The ".csv" file extension now determines the CSV format. We might do better by inspecting the first row of the data that comes in, but such a detection might be error-prone (I don't know yet). I would prefer to do this in a separate issue after this PR is merged.

Reply:
Is there any reliable way to detect line protocol on input in the existing code base? Or should we simply assume CSV data if input starts with an annotation comment (no real measurements probably start with #)

If the data input is supplied via stdin, the user should better let us know what data type (lp or csv) comes in ... the user knows, but the tool guesses. I could also imagine users supplying wrong data, from which we will then provide false advice about parsing errors.

@jsteenb2 Can you please comment on this?

Reply:
totally fine breakign this out into separate issue. Its a quality of life feature. I'm not advocating that we remove the format. I was thinking it coudl be optional if we have a file as an input.  I dont' think we have a reliable way to inspect a buffer and pull out if it is LP or csv 🤔. I could be wrong on that. Maybe our storage reviewer (@sebito91) would know.

Reply:
we do something similar to this in the pkg command. We just bet on the file extension being accurate and make decisions off of that if the `encoding` flag isn't set 👍 

https://github.com/influxdata/influxdb/blob/master/cmd/influx/pkg.go#L490

Reply:
Yes, the format flag is already optional herein, defaults to lp (backward compatibility) unless a file with .csv extension supplied.

https://github.com/influxdata/influxdb/blob/17003/writeFromCsv/cmd/influx/write.go#L154